File: drivers/src/i2c_interface.erl
%%%-------------------------------------------------------------------
%%% @author  drimtajm@github
%%% @copyright (C) 2013, Angela Johansson
%%% @doc
%%% An Erlang NIF that interfaces the I2C bus on the Raspberry Pi
%%%
%%% This interface uses the I2C ioctl driver which provides access to the
%%% I2C bus with a file descriptor. A precondition to using this software
%%% is a working I2C bus and an installed i2c-tools package.
%%%
%%% i2c_interface is free software:  you can redistribute it and/or modify
%%% it under the terms of the GNU Lesser General Public License as
%%% published by the Free Software Foundation, either version 3 of the
%%% License, or (at your option) any later version.
%%%
%%% i2c_interface is distributed in the hope that it will be useful,
%%% but WITHOUT ANY WARRANTY; without even the implied warranty of
%%% MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
%%% GNU Lesser General Public License for more details.
%%%
%%% You should have received a copy of the GNU Lesser General Public License
%%% along with this software.  If not, see <http://www.gnu.org/licenses/>.
%%%
%%% @end
%%% Created : 30 Jul 2013 by drimtajm
%%%-------------------------------------------------------------------
-module(i2c_interface).

-export([open_i2c_bus/1, close_i2c_bus/1]).
-export([read_i2c_byte/2, read_i2c_word/2]).
-export([write_i2c_byte/3, write_i2c_word/3]).

-define(nif_stub,
        erlang:nif_error({nif_not_loaded, module, ?MODULE, line, ?LINE})).

-on_load(on_load/0).

-type file_descriptor() :: integer().
-type i2c_address()     :: integer().
-type register()        :: integer().   %% TODO: constrain to [0; 255]
-type error_code()      :: atom().

on_load() ->
    PrivDir = case code:priv_dir(?MODULE) of
                  {error, bad_name} ->
                      EbinDir = filename:dirname(code:which(?MODULE)),
                      AppPath = filename:dirname(EbinDir),
                      filename:join(AppPath, "priv");
                  Path ->
                      Path
              end,
    Filename = filename:join(PrivDir, ?MODULE),
    ok = erlang:load_nif(Filename, 0).

-spec(open_i2c_bus(i2c_address()) -> {ok, file_descriptor()} | {error, error_code()}).
%%% @doc This opens the I2C bus, connects to the device at the specified
%%%      address and returns a file handle to the open bus.
open_i2c_bus(I2CAddress) ->
    open_i2c_bus_nif(I2CAddress).

-spec(close_i2c_bus(file_descriptor()) -> ok| {error, error_code()}).
%%% @doc This closes the I2C bus for a given file handle.
close_i2c_bus(FileHandle) ->
    close_i2c_bus_nif(FileHandle).

%% TODO: correct specification
-spec(read_i2c_byte(file_descriptor(), register()) -> {ok, integer()} | error).
%%% @doc This reads a register value from the I2C device.
read_i2c_byte(FileHandle, Register) ->
    read_i2c_byte_nif(FileHandle, Register).

%% TODO: correct specification
-spec(read_i2c_word(file_descriptor(), register()) -> {ok, integer()} | error).
%%% @doc This reads a register value from the I2C device.
read_i2c_word(FileHandle, Register) ->
    read_i2c_word_nif(FileHandle, Register).

write_i2c_byte(FileHandle, Register, Value) ->
    write_i2c_byte_nif(FileHandle, Register, Value).

write_i2c_word(FileHandle, Register, Value) ->
    write_i2c_word_nif(FileHandle, Register, Value).

%%%%%%%%%%%%%%%
%% Define stubs for NIF functions

open_i2c_bus_nif(_I2CAddress)                      -> ?nif_stub.
read_i2c_byte_nif(_FileHandle, _Register)          -> ?nif_stub.
read_i2c_word_nif(_FileHandle, _Register)          -> ?nif_stub.
write_i2c_byte_nif(_FileHandle, _Register, _Value) -> ?nif_stub.
write_i2c_word_nif(_FileHandle, _Register, _Value) -> ?nif_stub.
close_i2c_bus_nif(_FileHandle)                     -> ?nif_stub.

%%
%%%%%%%%%%%%%%%

File: drivers/c_src/i2c_interface.c
// An Erlang NIF which interfaces the i2c driver exposed by ioctl. 
// Copyright (C) 2013 Angela Johansson
//
// This file is part of free software: you can redistribute it and/or modify
// it under the terms of the GNU Lesser General Public License as
// published by the Free Software Foundation, either version 3 of the
// License, or (at your option) any later version.
//
// This i2c interface is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU Lesser General Public License for more details.
//
// You should have received a copy of the GNU Lesser General Public License
// along with this code.  If not, see <http://www.gnu.org/licenses/>.

#include <stdint.h>
#include <fcntl.h>
#include <unistd.h>
#include <errno.h>
#include <sys/ioctl.h>

#ifdef __ARM_EABI__
// ARM architecture, we're probably on a Raspberry Pi. Include "real" I2C headers
#include <linux/i2c-dev.h>
#else
// Not on ARM, probably compiling on a PC. Include I2C stub headers
#include "i2c-stub.h"
#endif

#include "erl_nif.h"

static ERL_NIF_TERM atom_ok;
static ERL_NIF_TERM atom_error;

static int load(ErlNifEnv* env, void** priv_data, ERL_NIF_TERM load_info) {
    atom_ok    = enif_make_atom(env, "ok");
    atom_error = enif_make_atom(env, "error");
    return 0;
}

static ERL_NIF_TERM errno2atom(ErlNifEnv *env, const int error_code) {
  switch (error_code) {
  case EACCES: return enif_make_atom(env, "eacces");
  case EINVAL: return enif_make_atom(env, "einval");
  case ENOENT: return enif_make_atom(env, "enoent");
  case EBADF:  return enif_make_atom(env, "ebadf");
    // TODO: implement other error codes
  default:     return enif_make_atom(env, "other");
  }
}

static ERL_NIF_TERM make_error(ErlNifEnv *env, const int error_code) {
  return enif_make_tuple(env, 2, atom_error, errno2atom(env, error_code));
}

static ERL_NIF_TERM open_i2c_bus_nif(ErlNifEnv *env, int argc,
				     const ERL_NIF_TERM argv[]) {
  int file, address, result;
  if (!enif_get_int(env, argv[0], &address)) {
    return enif_make_badarg(env);
  }
#ifdef __ARM_EABI__
  // My Raspberry Pi uses I2C bus 0.
  // If necessary, change the line to match the bus number of _your_ system:
  char *filename = "/dev/i2c-0";
#else
  char *filename = "i2c-bus-stub";
#endif
  if ((file = open(filename, O_RDWR)) < 0) {
    result = errno;
    return make_error(env, result);
  }
  // The bus is open, now communicate to the device pointed out by address
  if (ioctl(file, I2C_SLAVE, address) < 0) {
    return make_error(env, errno);
  }
  return enif_make_tuple(env, 2, atom_ok, enif_make_int(env, file));
}

static ERL_NIF_TERM close_i2c_bus_nif(ErlNifEnv *env, int argc,
				      const ERL_NIF_TERM argv[]) {
  int file, result;
  if (!enif_get_int(env, argv[0], &file)) {
    return enif_make_badarg(env);
  }
  if ((result = close(file)) < 0) {
    result = errno;
    return enif_make_tuple(env, 2, atom_error, errno2atom(env, result));
    //return enif_make_tuple(env, 2, atom_error, enif_make_int(env, result));  
  }
  return atom_ok;
}

static ErlNifFunc nif_funcs[] =
    {
        // setup
      {"open_i2c_bus_nif",            1, open_i2c_bus_nif},
      {"close_i2c_bus_nif",           1, close_i2c_bus_nif}
    };

ERL_NIF_INIT(i2c_interface, nif_funcs, load, NULL, NULL, NULL)
